### PR TITLE
Add eval_start_step flag and allow evaluation from step 0

### PR DESCRIPTION
**Pull request by zxhe-sean**
# Description

This PR adds full support for the `eval_start_step` configuration flag in MaxText to delay or control when interval-based evaluations trigger during pre-training, and updates the evaluation step check from `step > start_step` to `step >= start_step`.

### Why is this change being made

1. **Support for `eval_start_step`:** The need of the delay is coming from the new validation [requirement of MLPerf](https://github.com/mlcommons/training/pull/891)

> The validation starts after N(GBS) = GBS*FLOOR(42+24576/GBS) samples - this gives us 2-3 spare evaluations before we hit the RCP min, and ~4 before we hit the RCP avg. 

2. **Support evaluation on step 0 / `start_step` (`step >= start_step`):** The evaluation condition previously checked `step > start_step`, strictly preventing evaluations from firing on the very first step (`step == start_step`), such as when `start_step = 0` and `eval_interval = 1`. Updating this to `step >= start_step` allows evaluation to fire on step 0 or the starting step when configured.

FIXES: b/535641549

# Tests

* When setting `eval_start_step=10`, the log shows the eval started at step 10: https://fusion2.corp.google.com/invocations/f6ea1094-7d28-4c76-be75-f0bbaeb9ac5d/log

* The eval now can start from step 0: https://fusion2.corp.google.com/invocations/7cb76b6f-e36e-4ba7-ab17-aa90a9e1c6e6/log

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/base.yml [hunk 1]
```diff
@@ -983,6 +983,7 @@ decode_sampling_nucleus_p: -1 # set if you're doing nucleus / top-p
 decode_sampling_top_k: 0 # set if you're doing top-k
 decode_sampling_temperature: 1.
 
+eval_start_step: -1 # start eval after train step is greater than eval_start_step, for MLPerf, we want it to be FLOOR(42+24576/GBS)
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml]: Let's remove `for MLPerf, we want it to be FLOOR(42+24576/GBS)`

Nit to be concise: `# Start evaluation once train_step >= eval_start_step`

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -1573,6 +1573,10 @@ class TrainingLoop(BaseModel):
       description="Total number of training steps. -1 defaults to learning_rate_schedule_steps.",
   )
   log_period: int = Field(100, description="Frequency (in steps) to log metrics and flush Tensorboard.")
+  eval_start_step: int = Field(
+      -1,
+      description="Start evaluation after training step is greater than eval_start_step.",
```
Comment (RissyRan) [inline src/maxtext/configs/types.py]: To be aligned above if any changes.

File: src/maxtext/trainers/pre_train/train.py [hunk 3]
```diff
@@ -712,7 +713,7 @@ def training_loop_iteration(
         all_host_upload=dump_hlo_upload_all,
     )
 
-  if eval_interval > 0 and step > start_step and (step + 1) % eval_interval == 0:
+  if eval_interval > 0 and step > start_step and step > eval_start_step and (step + 1) % eval_interval == 0:
```
Comment (RissyRan) [inline src/maxtext/trainers/pre_train/train.py]: Please cross check if it's > vs. >=. For instance, if we set up train step as 100, it will train from step 0 to step 99. Cross check which `step` it is, and see if `>` or `>=` applies

File: tests/unit/pyconfig_test.py [hunk 4]
```diff
@@ -317,6 +317,21 @@ def test_local_sa_flags_explicit_override(self):
     self.assertEqual(config.local_sa_v_layout, "SEQ_MINOR")
     self.assertFalse(config.local_use_splash_scheduler)
 
+  def test_eval_start_step_config(self):
```
Comment (RissyRan) [inline tests/unit/pyconfig_test.py:320]: Could you also add your end-to-end check in the PR description?

**Review by RissyRan — COMMENTED** [review]
Thanks for the change!

File: src/maxtext/configs/base.yml (hunk 1 above)
Comment (zxhe-sean) [reply src/maxtext/configs/base.yml]: Done

File: src/maxtext/trainers/pre_train/train.py (hunk 3 above)
Comment (zxhe-sean) [reply src/maxtext/trainers/pre_train/train.py]: After offline discussion with Ran, we've agreed that when we set `eval_start_step=10`, it means we need to start eval after the training has gone 10 steps. Therefore, we need to start eval at the 11th step, which is step 10.   Change `>` to `>=`

File: tests/unit/pyconfig_test.py (hunk 4 above)
Comment (zxhe-sean) [reply tests/unit/pyconfig_test.py:320]: added my end-to-end check in the PR with `eval_start_step=10`

**Review by aireenmei — APPROVED** [review]
LGTM!

**Review by RissyRan — APPROVED** [review]
